File: snake/cpp/src/Renderer4D.cpp
#include "Renderer4D.h"
#include <cmath>

Renderer4D::Renderer4D() : Renderer3D() {}

void Renderer4D::recordGhostTrail(ContinuousSnake& snake) {
    if (GetRandomValue(0, 100) > 5) return;
    auto segs = snake.getSegments();
    GhostTrail trail;
    trail.segments = segs;
    trail.time = time4d;
    trail.color1 = snake.stage().color1;
    trail.color2 = snake.stage().color2;
    trail.radius = snake.stage().radius;
    ghostTrails.push_back(trail);
    if ((int)ghostTrails.size() > maxGhostTrails) {
        ghostTrails.erase(ghostTrails.begin());
    }
}

void Renderer4D::drawGhostTrails3D() {
    for (auto& trail : ghostTrails) {
        float age = time4d - trail.time;
        float alpha = std::max(0.f, 0.4f - age * 0.1f);
        if (alpha <= 0) continue;

        unsigned char a = (unsigned char)(alpha * 255);
        for (int i = 0; i < (int)trail.segments.size(); i += 2) {
            float t = 1.f - ((float)i / (float)trail.segments.size());
            float r = trail.radius * 2.f * (0.4f + t * 0.6f);
            float bob = sinf(time4d * 5 + i) * 10.f;
            DrawSphere({trail.segments[i].x, r + bob, trail.segments[i].y}, r,
                       {trail.color1.r, trail.color1.g, trail.color1.b, a});
        }
    }
}

void Renderer4D::render(ContinuousSnake& snake, FoodManager& food, InfiniteArena& arena, StoryManager& story, int sw, int sh) {
    time4d += GetFrameTime();
    time = time4d; // sync parent time

    recordGhostTrail(snake);
    updateCamera(snake);

    // Breathing camera
    float camDist = 500 + sinf(time4d) * 100;
    float camHeight = 450 + cosf(time4d) * 50;
    float dX = snake.x - cosf(snake.angle) * camDist;
    float dZ = snake.y - sinf(snake.angle) * camDist;
    camera.position = {dX, camHeight, dZ};

    ClearBackground({8, 0, 32, 255});

    BeginMode3D(camera);
        drawGrid3D(snake.x, snake.y);
        drawGhostTrails3D();
        drawFood3D(food);
        drawBoss3D(story);
        drawSnake3D(snake);
    EndMode3D();

    story.drawCinematic(sw, sh);

    // Temporal HUD label
    DrawText("[ TEMPORAL DIMENSION ]", sw/2 - 120, sh - 40, 20, {180, 74, 255, 150});
}

File: src/renderers/Renderer4D.js
// ============================================
// 4D RENDERER — Infinite Time-Warp Dimension
// ============================================

import * as THREE from 'three';
import { Renderer3D } from './Renderer3D.js';

export class Renderer4D extends Renderer3D {
  constructor(container) {
    super(container);
    this.ghostTrails = [];
    this.ghostMeshes = [];
    this.maxGhostTrails = 30;
  }

  init(arenaSize = 10000) {
    super.init(arenaSize);

    this.scene.fog = new THREE.FogExp2(0x080020, 0.002);
    const temporalLight = new THREE.AmbientLight(0x4400aa, 0.5);
    this.scene.add(temporalLight);
  }

  recordGhostTrail(snake) {
    if (Math.random() > 0.05) return; // Only snapshot rarely
    const segs = snake.getSegments().map(s => ({ x: s.x, y: s.y }));
    this.ghostTrails.push({
      segments: segs,
      time: this.time,
      color1: snake.stage.color1,
      color2: snake.stage.color2,
      radius: snake.stage.radius
    });
    if (this.ghostTrails.length > this.maxGhostTrails) {
      this.ghostTrails.shift();
    }
  }

  updateGhostTrails() {
    this.ghostMeshes.forEach(m => {
      this.scene.remove(m);
      if (m.geometry) m.geometry.dispose();
      if (m.material) m.material.dispose();
    });
    this.ghostMeshes = [];

    this.ghostTrails.forEach(trail => {
      const age = (this.time - trail.time);
      const alpha = Math.max(0, 0.4 - age * 0.1);
      if (alpha <= 0) return;

      trail.segments.forEach((seg, i) => {
        if (i % 2 !== 0) return; 
        const t = 1 - (i / trail.segments.length);
        const radius = trail.radius * 2 * (0.4 + t * 0.6);

        const geo = new THREE.SphereGeometry(radius, 8, 8);
        const mat = new THREE.MeshStandardMaterial({
          color: new THREE.Color(trail.color1),
          emissive: new THREE.Color(trail.color2),
          emissiveIntensity: 0.5,
          transparent: true,
          opacity: alpha,
          metalness: 1, roughness: 0
        });
        const mesh = new THREE.Mesh(geo, mat);
        mesh.position.set(seg.x, radius + Math.sin(this.time * 5 + i) * 10, seg.y);
        this.scene.add(mesh);
        this.ghostMeshes.push(mesh);
      });
    });
  }

  render(snake, food, arena, story) {
    if (!this.initialized) return;
    this.time += this.clock.getDelta();

    this.recordGhostTrail(snake);
    this.updateGhostTrails();

    this.updateSnake(snake);
    this.updateFood(food);
    this.updateStory(story);

    // Dynamic Camera
    const lookDist = 150;
    const targetX = snake.x + Math.cos(snake.angle) * lookDist;
    const targetZ = snake.y + Math.sin(snake.angle) * lookDist;

    const camDist = 500 + Math.sin(this.time)*100; // breathing camera
    const camHeight = 450 + Math.cos(this.time)*50;
    
    // Lerp camera behind snake
    const dX = snake.x - Math.cos(snake.angle) * camDist;
    const dZ = snake.y - Math.sin(snake.angle) * camDist;

    this.camera.position.set(dX, camHeight, dZ);
    this.camera.lookAt(targetX, 0, targetZ);

    // Hue shift
    const hue = (this.time * 0.05) % 1;
    this.scene.fog.color.setHSL(hue, 0.8, 0.1);

    this.renderer.render(this.scene, this.camera);
  }
}
